### DISCUSSION TITLE
BentoML KMeans error: Exception happened in API function: buffer source array is read-only

### DISCUSSION THREAD

**Discussion by mat-berna** [bentoml/BentoML#1217] [Q&A] [answered]
Hi,

I'm trying to deploy a simple KMeans algorithm using BentoML.
I create a conda enviroment `bentoml_test` with the following comands:

```
conda create --name bentoml_test python=3.8.3
conda activate bentoml_test
pip install --trusted-host files.pythonhosted.org --trusted-host pypi.org --trusted-host pypi.python.org bentoml==0.9.0
conda install scikit-learn==0.23.1
conda install pandas==1.0.5
```

The following script creates a simple IrisClassifier image using a KMeans algorithm:

```
from sklearn import svm
from sklearn import datasets
from sklearn.cluster import KMeans 

from iris_classifier import IrisClassifier 

if __name__ == "__main__":
    iris = datasets.load_iris()
    x = iris.data
    y = iris.target
    
    km =  KMeans(n_clusters=2, random_state=0).fit(x)

    iris_classifier_service = IrisClassifier() 

    iris_classifier_service.pack('model', km) 

    saved_path = iris_classifier_service.save() 
```

And the serving is ok:

```
bentoml serve IrisClassifier:latest
INFO - Getting latest version IrisClassifier:20201030125549_C3FF42
INFO - Starting BentoML API server in development mode..
 * Serving Flask app "IrisClassifier" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
```

But when I call `/predict` method on Swagger interface I have the following error:

![image](https://user-images.githubusercontent.com/9885388/97720639-cc66c280-1ac8-11eb-8e51-c75e54b81559.png)

Can someone help me?
Thanks.

**Comment by yubozhao** [comment, marked as answer]
@berna4 Thanks for reporting

This error is due to how BentoML is loading SKLearn model with `joblib` library.  Right now, BentoML is loading with `mmap_mode` to `r`, and that caused the `read-only` issue.

Since Sklearn model can be persist with pickle, we can use the `PickleArtifacts`. For now you can change the `iris_classifier.py` with the `bentoml.service.artifacts.common.PickleArtifact`

```python
# iris_clasifier.py
...
# from bentoml.frameworks.sklearn import SklearnModelArtifact
from bentoml.service.artifact.common import PIckleArtifact
...

@artifacts([PickleArtifact('model')])
class IrisClassifier(BentoService):
    @api(...)
    def predict(self, df):
        ...
```

**Reply by mat-berna** [reply]
@yubozhao Thanks for the answer.
I tried the solution and verified that it works correctly.
